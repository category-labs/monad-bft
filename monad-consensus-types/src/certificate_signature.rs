use monad_crypto::{
    secp256k1::{Error as SecpError, KeyPair as SecpKeyPair, PubKey as SecpPubKey, SecpSignature},
    NopSignature,
};

use crate::{message_signature::MessageSignature, validation::Hashable};

pub trait CertificateKeyPair: Send + Sized + Sync + 'static {
    type PubKeyType: std::cmp::Eq + std::hash::Hash + Copy;
    type Error: std::error::Error + Send + Sync;

    fn from_bytes(secret: impl AsMut<[u8]>) -> Result<Self, Self::Error>;
    fn pubkey(&self) -> Self::PubKeyType;
}

pub trait CertificateSignature:
    Copy + Clone + Eq + Hashable + Send + Sync + std::fmt::Debug + 'static
{
    type KeyPairType: CertificateKeyPair;
    type Error: std::error::Error + Send + Sync;

    fn sign(msg: &[u8], keypair: &Self::KeyPairType) -> Self;
    fn verify(
        &self,
        msg: &[u8],
        pubkey: &<Self::KeyPairType as CertificateKeyPair>::PubKeyType,
    ) -> Result<(), Self::Error>;

    fn serialize(&self) -> Vec<u8>;
    fn deserialize(signature: &[u8]) -> Result<Self, Self::Error>;
}

pub trait CertificateSignatureRecoverable: CertificateSignature {
    fn recover_pubkey(
        &self,
        msg: &[u8],
    ) -> Result<
        <<Self as CertificateSignature>::KeyPairType as CertificateKeyPair>::PubKeyType,
        <Self as CertificateSignature>::Error,
    >;
}

impl CertificateKeyPair for SecpKeyPair {
    type PubKeyType = SecpPubKey;
    type Error = SecpError;

    fn from_bytes(secret: impl AsMut<[u8]>) -> Result<Self, Self::Error> {
        Self::from_bytes(secret)
    }

    fn pubkey(&self) -> Self::PubKeyType {
        self.pubkey()
    }
}

impl Hashable for SecpSignature {
    fn hash<H: crate::validation::Hasher>(&self, state: &mut H) {
        let slice = unsafe { std::mem::transmute::<Self, [u8; 65]>(*self) };
        state.update(slice)
    }
}

impl CertificateSignature for SecpSignature {
    type KeyPairType = SecpKeyPair;
    type Error = SecpError;

    fn sign(msg: &[u8], keypair: &Self::KeyPairType) -> Self {
        keypair.sign(msg)
    }

    fn verify(
        &self,
        msg: &[u8],
        pubkey: &<Self::KeyPairType as CertificateKeyPair>::PubKeyType,
    ) -> Result<(), Self::Error> {
        pubkey.verify(msg, self)
    }

    fn serialize(&self) -> Vec<u8> {
        self.serialize().to_vec()
    }

    fn deserialize(signature: &[u8]) -> Result<Self, Self::Error> {
        Self::deserialize(signature)
    }
}

impl CertificateSignatureRecoverable for SecpSignature {
    fn recover_pubkey(
        &self,
        msg: &[u8],
    ) -> Result<
        <<Self as CertificateSignature>::KeyPairType as CertificateKeyPair>::PubKeyType,
        <Self as CertificateSignature>::Error,
    > {
        self.recover_pubkey(msg)
    }
}

impl Hashable for NopSignature {
    fn hash<H: crate::validation::Hasher>(&self, state: &mut H) {
        let slice = unsafe { std::mem::transmute::<Self, [u8; 72]>(*self) };
        state.update(slice)
    }
}

impl CertificateSignature for NopSignature {
    type KeyPairType = SecpKeyPair;
    type Error = SecpError;

    fn sign(msg: &[u8], keypair: &Self::KeyPairType) -> Self {
        <Self as MessageSignature>::sign(msg, keypair)
    }

    fn verify(
        &self,
        msg: &[u8],
        pubkey: &<Self::KeyPairType as CertificateKeyPair>::PubKeyType,
    ) -> Result<(), Self::Error> {
        <Self as MessageSignature>::verify(self, msg, pubkey)
    }

    fn serialize(&self) -> Vec<u8> {
        <Self as MessageSignature>::serialize(self)
    }

    fn deserialize(signature: &[u8]) -> Result<Self, Self::Error> {
        <Self as MessageSignature>::deserialize(signature)
    }
}

impl CertificateSignatureRecoverable for NopSignature {
    fn recover_pubkey(
        &self,
        msg: &[u8],
    ) -> Result<
        <<Self as CertificateSignature>::KeyPairType as CertificateKeyPair>::PubKeyType,
        <Self as CertificateSignature>::Error,
    > {
        <Self as MessageSignature>::recover_pubkey(self, msg)
    }
}

#[cfg(test)]
mod test {
    use std::ops::AddAssign;

    use monad_crypto::secp256k1::SecpSignature;

    // valid certificate signature tests
    use crate::certificate_signature::{
        CertificateKeyPair, CertificateSignature, CertificateSignatureRecoverable,
    };
    macro_rules! test_all_certificate_signature {
        ($test_name:ident, $test_code:block) => {
            mod $test_name {
                use monad_crypto::{secp256k1::SecpSignature, NopSignature};

                use super::*;

                fn invoke<T>()
                where
                    T: CertificateSignature + std::fmt::Debug,
                    <T::KeyPairType as CertificateKeyPair>::PubKeyType:
                        std::cmp::Eq + std::fmt::Debug,
                {
                    $test_code
                }

                #[test]
                fn secpsignature() {
                    invoke::<SecpSignature>();
                }

                #[test]
                fn nopsignature() {
                    invoke::<NopSignature>();
                }

                // TODO: add module for bls signature
            }
        };
    }

    macro_rules! test_all_certificate_signature_recoverable {
        ($test_name:ident, $test_code:block) => {
            mod $test_name {
                use monad_crypto::{secp256k1::SecpSignature, NopSignature};

                use super::*;

                fn invoke<T>()
                where
                    T: CertificateSignatureRecoverable + std::fmt::Debug,
                    <T::KeyPairType as CertificateKeyPair>::PubKeyType:
                        std::cmp::Eq + std::fmt::Debug,
                {
                    $test_code
                }

                #[test]
                fn secpsignature() {
                    invoke::<SecpSignature>();
                }

                #[test]
                fn nopsignature() {
                    invoke::<NopSignature>();
                }
            }
        };
    }

    test_all_certificate_signature!(test_keypair_deterministic_creation, {
        let mut s1 = [127_u8; 32];
        let mut s2 = [127_u8; 32];

        assert_eq!(s1, s2);

        let k1 = <T::KeyPairType as CertificateKeyPair>::from_bytes(s1.as_mut_slice()).unwrap();
        let k2 = <T::KeyPairType as CertificateKeyPair>::from_bytes(s2.as_mut_slice()).unwrap();

        assert_eq!(k1.pubkey(), k2.pubkey());
    });

    test_all_certificate_signature!(test_serialization_roundtrip, {
        let mut s = [127_u8; 32];
        let certkey = <T::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let sig = T::sign(msg, &certkey);

        let sig_bytes = sig.serialize();
        let sig_de = T::deserialize(sig_bytes.as_ref()).unwrap();

        assert_eq!(sig, sig_de);
    });

    test_all_certificate_signature!(test_signature_verify, {
        let mut s = [127_u8; 32];
        let certkey = <T::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let sig = T::sign(msg, &certkey);

        assert!(sig.verify(msg, &certkey.pubkey()).is_ok());
    });

    test_all_certificate_signature_recoverable!(test_recover, {
        let mut s = [127_u8; 32];
        let certkey = <T::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let sig = T::sign(msg, &certkey);

        assert_eq!(sig.recover_pubkey(msg).unwrap(), certkey.pubkey());
    });

    // invalid certificate signature tests
    #[test]
    fn test_verify_error() {
        let mut s = [127_u8; 32];
        let certkey = <<SecpSignature as CertificateSignature>::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let invalid_msg = b"bye world";
        let sig = <SecpSignature as CertificateSignature>::sign(msg, &certkey);

        assert!(<SecpSignature as CertificateSignature>::verify(
            &sig,
            invalid_msg,
            &certkey.pubkey()
        )
        .is_err());
    }

    #[test]
    fn test_deser_error() {
        let mut s = [127_u8; 32];
        let certkey = <<SecpSignature as CertificateSignature>::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let sig = <SecpSignature as CertificateSignature>::sign(msg, &certkey);

        let mut sig_bytes = <SecpSignature as CertificateSignature>::serialize(&sig);

        // the last byte is the recoveryId
        // recoveryId is 0..=3, adding 4 makes it invalid
        sig_bytes.last_mut().unwrap().add_assign(5);

        assert!(<SecpSignature as CertificateSignature>::deserialize(&sig_bytes).is_err());
    }
}
