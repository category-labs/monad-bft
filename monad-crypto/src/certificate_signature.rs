use std::{
    collections::hash_map::DefaultHasher,
    fmt::{Debug, Display},
    hash::{Hash, Hasher},
};

use crate::{
    bls12_381::{BlsError, BlsKeyPair, BlsPubKey, BlsSignature},
    hasher::Hashable,
    secp256k1::{Error as SecpError, KeyPair as SecpKeyPair, PubKey as SecpPubKey, SecpSignature},
    NopKeyPair, NopPubKey, NopSignature,
};

pub trait PubKey:
    Debug + Eq + Hash + Ord + PartialOrd + Copy + Send + Sync + Unpin + 'static
{
    type Error: Display + Debug + Send + Sync;
    fn from_bytes(pubkey: &[u8]) -> Result<Self, Self::Error>;
    fn bytes(&self) -> Vec<u8>;
}

pub trait CertificateKeyPair: Send + Sized + Sync + 'static {
    type PubKeyType: PubKey;
    type Error: Display + Debug + Send + Sync;

    fn from_bytes(secret: &mut [u8]) -> Result<Self, Self::Error>;
    fn pubkey(&self) -> Self::PubKeyType;
}

pub type CertificateSignaturePubKey<T> =
    <<T as CertificateSignature>::KeyPairType as CertificateKeyPair>::PubKeyType;

pub trait CertificateSignature:
    Copy + Clone + Eq + Hashable + Send + Sync + std::fmt::Debug + std::hash::Hash + 'static
{
    type KeyPairType: CertificateKeyPair;
    type Error: Display + Debug + Send + Sync;

    fn sign(msg: &[u8], keypair: &Self::KeyPairType) -> Self;
    fn verify(
        &self,
        msg: &[u8],
        pubkey: &CertificateSignaturePubKey<Self>,
    ) -> Result<(), Self::Error>;

    fn serialize(&self) -> Vec<u8>;
    fn deserialize(signature: &[u8]) -> Result<Self, Self::Error>;
}

pub trait CertificateSignatureRecoverable: CertificateSignature {
    fn recover_pubkey(
        &self,
        msg: &[u8],
    ) -> Result<CertificateSignaturePubKey<Self>, <Self as CertificateSignature>::Error>;
}

impl PubKey for SecpPubKey {
    type Error = SecpError;

    fn from_bytes(pubkey: &[u8]) -> Result<Self, Self::Error> {
        Self::from_slice(pubkey)
    }

    fn bytes(&self) -> Vec<u8> {
        SecpPubKey::bytes(self)
    }
}

impl CertificateKeyPair for SecpKeyPair {
    type PubKeyType = SecpPubKey;
    type Error = SecpError;

    fn from_bytes(secret: &mut [u8]) -> Result<Self, Self::Error> {
        Self::from_bytes(secret)
    }

    fn pubkey(&self) -> Self::PubKeyType {
        self.pubkey()
    }
}

impl CertificateSignature for SecpSignature {
    type KeyPairType = SecpKeyPair;
    type Error = SecpError;

    fn sign(msg: &[u8], keypair: &Self::KeyPairType) -> Self {
        keypair.sign(msg)
    }

    fn verify(
        &self,
        msg: &[u8],
        pubkey: &CertificateSignaturePubKey<Self>,
    ) -> Result<(), Self::Error> {
        pubkey.verify(msg, self)
    }

    fn serialize(&self) -> Vec<u8> {
        self.serialize().to_vec()
    }

    fn deserialize(signature: &[u8]) -> Result<Self, Self::Error> {
        Self::deserialize(signature)
    }
}

impl PubKey for NopPubKey {
    type Error = &'static str;

    fn from_bytes(pubkey: &[u8]) -> Result<Self, Self::Error> {
        Ok(Self(
            pubkey
                .try_into()
                .map_err(|_| "couldn't deserialize pubkey")?,
        ))
    }

    fn bytes(&self) -> Vec<u8> {
        self.0.to_vec()
    }
}

impl CertificateKeyPair for NopKeyPair {
    type PubKeyType = NopPubKey;
    type Error = &'static str;

    fn from_bytes(secret: &mut [u8]) -> Result<Self, Self::Error> {
        Ok(Self {
            pubkey: NopPubKey::from_bytes(secret)?,
        })
    }

    fn pubkey(&self) -> Self::PubKeyType {
        self.pubkey
    }
}

impl CertificateSignatureRecoverable for SecpSignature {
    fn recover_pubkey(
        &self,
        msg: &[u8],
    ) -> Result<CertificateSignaturePubKey<Self>, <Self as CertificateSignature>::Error> {
        self.recover_pubkey(msg)
    }
}

impl CertificateSignature for NopSignature {
    type KeyPairType = NopKeyPair;
    type Error = &'static str;

    fn sign(msg: &[u8], keypair: &Self::KeyPairType) -> Self {
        let mut hasher = DefaultHasher::new();
        hasher.write(msg);

        NopSignature {
            pubkey: keypair.pubkey,
            id: hasher.finish(),
        }
    }

    fn verify(
        &self,
        _msg: &[u8],
        pubkey: &CertificateSignaturePubKey<Self>,
    ) -> Result<(), Self::Error> {
        if &self.pubkey == pubkey {
            Ok(())
        } else {
            Err("invalid pubkey")
        }
    }

    fn serialize(&self) -> Vec<u8> {
        self.id
            .to_le_bytes()
            .into_iter()
            .chain(self.pubkey.bytes())
            .collect()
    }

    fn deserialize(signature: &[u8]) -> Result<Self, Self::Error> {
        let id = u64::from_le_bytes(signature[..8].try_into().unwrap());
        let pubkey = NopPubKey::from_bytes(&signature[8..])?;
        Ok(Self { pubkey, id })
    }
}

impl CertificateSignatureRecoverable for NopSignature {
    fn recover_pubkey(
        &self,
        _msg: &[u8],
    ) -> Result<CertificateSignaturePubKey<Self>, <Self as CertificateSignature>::Error> {
        Ok(self.pubkey)
    }
}

impl PubKey for BlsPubKey {
    type Error = BlsError;

    fn from_bytes(pubkey: &[u8]) -> Result<Self, Self::Error> {
        Self::deserialize(pubkey)
    }

    fn bytes(&self) -> Vec<u8> {
        self.serialize()
    }
}

impl CertificateKeyPair for BlsKeyPair {
    type PubKeyType = BlsPubKey;
    type Error = BlsError;

    fn from_bytes(secret: &mut [u8]) -> Result<Self, Self::Error> {
        Self::from_bytes(secret)
    }

    fn pubkey(&self) -> Self::PubKeyType {
        self.pubkey()
    }
}

impl CertificateSignature for BlsSignature {
    type KeyPairType = BlsKeyPair;
    type Error = BlsError;

    fn sign(msg: &[u8], keypair: &Self::KeyPairType) -> Self {
        keypair.sign(msg)
    }

    fn verify(
        &self,
        msg: &[u8],
        pubkey: &CertificateSignaturePubKey<Self>,
    ) -> Result<(), Self::Error> {
        self.verify(msg, pubkey)
    }

    fn serialize(&self) -> Vec<u8> {
        self.serialize()
    }

    fn deserialize(signature: &[u8]) -> Result<Self, Self::Error> {
        Self::deserialize(signature)
    }
}

#[cfg(test)]
mod test {
    use std::ops::AddAssign;

    // valid certificate signature tests
    use crate::certificate_signature::{
        CertificateKeyPair, CertificateSignature, CertificateSignaturePubKey,
        CertificateSignatureRecoverable,
    };
    use crate::{bls12_381::BlsSignature, secp256k1::SecpSignature};
    macro_rules! test_all_certificate_signature {
        ($test_name:ident, $test_code:block) => {
            mod $test_name {
                use super::*;
                use crate::{secp256k1::SecpSignature, NopSignature};

                fn invoke<T>()
                where
                    T: CertificateSignature + std::fmt::Debug,
                    CertificateSignaturePubKey<T>: std::cmp::Eq + std::fmt::Debug,
                {
                    $test_code
                }

                #[test]
                fn secpsignature() {
                    invoke::<SecpSignature>();
                }

                #[test]
                fn nopsignature() {
                    invoke::<NopSignature>();
                }

                #[test]
                fn blssignature() {
                    invoke::<BlsSignature>();
                }
            }
        };
    }

    macro_rules! test_all_certificate_signature_recoverable {
        ($test_name:ident, $test_code:block) => {
            mod $test_name {
                use super::*;
                use crate::{secp256k1::SecpSignature, NopSignature};

                fn invoke<T>()
                where
                    T: CertificateSignatureRecoverable + std::fmt::Debug,
                    CertificateSignaturePubKey<T>: std::cmp::Eq + std::fmt::Debug,
                {
                    $test_code
                }

                #[test]
                fn secpsignature() {
                    invoke::<SecpSignature>();
                }

                #[test]
                fn nopsignature() {
                    invoke::<NopSignature>();
                }
            }
        };
    }

    test_all_certificate_signature!(test_keypair_deterministic_creation, {
        let mut s1 = [127_u8; 32];
        let mut s2 = [127_u8; 32];

        assert_eq!(s1, s2);

        let k1 = <T::KeyPairType as CertificateKeyPair>::from_bytes(s1.as_mut_slice()).unwrap();
        let k2 = <T::KeyPairType as CertificateKeyPair>::from_bytes(s2.as_mut_slice()).unwrap();

        assert_eq!(k1.pubkey(), k2.pubkey());
    });

    test_all_certificate_signature!(test_serialization_roundtrip, {
        let mut s = [127_u8; 32];
        let certkey = <T::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let sig = T::sign(msg, &certkey);

        let sig_bytes = sig.serialize();
        let sig_de = T::deserialize(sig_bytes.as_ref()).unwrap();

        assert_eq!(sig, sig_de);
    });

    test_all_certificate_signature!(test_signature_verify, {
        let mut s = [127_u8; 32];
        let certkey = <T::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let sig = T::sign(msg, &certkey);

        assert!(sig.verify(msg, &certkey.pubkey()).is_ok());
    });

    test_all_certificate_signature_recoverable!(test_recover, {
        let mut s = [127_u8; 32];
        let certkey = <T::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let sig = T::sign(msg, &certkey);

        assert_eq!(sig.recover_pubkey(msg).unwrap(), certkey.pubkey());
    });

    // invalid certificate signature tests
    #[test]
    fn test_verify_error() {
        let mut s = [127_u8; 32];
        let certkey = <<SecpSignature as CertificateSignature>::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let invalid_msg = b"bye world";
        let sig = <SecpSignature as CertificateSignature>::sign(msg, &certkey);

        assert!(<SecpSignature as CertificateSignature>::verify(
            &sig,
            invalid_msg,
            &certkey.pubkey()
        )
        .is_err());
    }

    #[test]
    fn test_deser_error() {
        let mut s = [127_u8; 32];
        let certkey = <<SecpSignature as CertificateSignature>::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let sig = <SecpSignature as CertificateSignature>::sign(msg, &certkey);

        let mut sig_bytes = <SecpSignature as CertificateSignature>::serialize(&sig);

        // the last byte is the recoveryId
        // recoveryId is 0..=3, adding 4 makes it invalid
        sig_bytes.last_mut().unwrap().add_assign(5);

        assert!(<SecpSignature as CertificateSignature>::deserialize(&sig_bytes).is_err());
    }
}
