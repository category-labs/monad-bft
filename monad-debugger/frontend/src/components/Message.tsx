import { Component } from "solid-js";
import { MessageFragment } from "../generated/graphql";

const Message: Component<{
    message: MessageFragment,
}> = (props) => {
    const message = () => props.message;

    if (message().__typename !== "GraphQLConsensusMessage") {
        return <div>Unknown message</div>
    }
    const round = () => message().round;

    const messageType = message().message;
    switch (messageType.__typename) {
        case "GraphQLProposal":
            const seqNum = messageType.seqNum;
            return (
                <div class="p-2 lg:p-4 text-nowrap rounded-2xl bg-green-400/50 border border-black shadow-sm">
                    <div class="text-xl lg:text-2xl">Proposal</div>
                    <div>
                        <span class="hidden lg:inline">round=</span>
                        <span class="inline lg:hidden">r=</span>
                        {round()}
                    </div>
                    <div>
                        <span class="hidden lg:inline">seq_num=</span>
                        <span class="inline lg:hidden">n=</span>
                        {seqNum}
                    </div>
                </div>
            );
        case "GraphQLVote":
            return (
                <div class="p-2 lg:p-4 text-nowrap rounded-2xl bg-green-400/15 border border-black shadow-sm">
                    <div class="text-xl lg:text-2xl">Vote</div>
                    <div>
                        <span class="hidden lg:inline">round=</span>
                        <span class="inline lg:hidden">r=</span>
                        {round()}
                    </div>
                </div>
            );
        case "GraphQLTimeout":
            return (
                <div class="p-2 lg:p-4 text-nowrap rounded-2xl bg-red-400/25 border border-black shadow-sm">
                    <div class="text-xl lg:text-2xl">Timeout</div>
                    <div>
                        <span class="hidden lg:inline">round=</span>
                        <span class="inline lg:hidden">r=</span>
                        {round()}
                    </div>
                </div>
            );
        case "GraphQLRoundRecovery":
            return (
                <div class="p-2 lg:p-4 text-nowrap rounded-2xl border border-black shadow-sm">
                    <div class="text-xl lg:text-2xl">RoundRecovery</div>
                    <div>
                        <span class="hidden lg:inline">round=</span>
                        <span class="inline lg:hidden">r=</span>
                        {round()}
                    </div>
                </div>
            );
        case "GraphQLNoEndorsement":
            return (
                <div class="p-2 lg:p-4 text-nowrap rounded-2xl border border-black shadow-sm">
                    <div class="text-xl lg:text-2xl">NoEndorsement</div>
                    <div>
                        <span class="hidden lg:inline">round=</span>
                        <span class="inline lg:hidden">r=</span>
                        {round()}
                    </div>
                </div>
            );
        default:
            assertUnreachable(messageType);
            return <div>Unknown message type</div>;
    }
};

function assertUnreachable(x?: never): never {
    throw new Error();
}

export default Message;
