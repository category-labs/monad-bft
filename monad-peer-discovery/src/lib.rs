use std::{
    collections::HashMap,
    net::{Ipv4Addr, SocketAddrV4},
    time::Duration,
};

use alloy_rlp::{Decodable, Encodable, encode_list};
use monad_crypto::certificate_signature::{
    CertificateSignature, CertificateSignaturePubKey, CertificateSignatureRecoverable,
};
use monad_executor_glue::Message;
use monad_types::NodeId;
use tracing::warn;

pub mod discovery;
pub mod mock;

pub type PeerDiscMetrics = HashMap<&'static str, u64>;

#[derive(Debug, Clone, Copy)]
pub struct Ping {
    id: u32,
    local_record_seq: u64,
}

#[derive(Debug, Clone, Copy)]
pub struct Pong {
    ping_id: u32,
    local_record_seq: u64,
}

#[derive(Debug, Clone)]
pub struct PeerLookupRequest<ST: CertificateSignatureRecoverable> {
    lookup_id: u32,
    target: NodeId<CertificateSignaturePubKey<ST>>,
}

#[derive(Debug, Clone)]
pub struct PeerLookupResponse<ST: CertificateSignatureRecoverable> {
    lookup_id: u32,
    target: NodeId<CertificateSignaturePubKey<ST>>,
    name_records: Vec<MonadNameRecord<ST>>,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct NameRecord {
    pub address: SocketAddrV4,
    pub seq: u64,
}

impl Encodable for NameRecord {
    fn encode(&self, out: &mut dyn alloy_rlp::BufMut) {
        let enc: [&dyn Encodable; 3] =
            [&self.address.ip().octets(), &self.address.port(), &self.seq];
        encode_list::<_, dyn Encodable>(&enc, out);
    }
}

impl Decodable for NameRecord {
    fn decode(buf: &mut &[u8]) -> alloy_rlp::Result<Self> {
        let buf = &mut alloy_rlp::Header::decode_bytes(buf, true)?;

        let Ok(ip) = <[u8; 4]>::decode(buf) else {
            warn!("ip address decode failed: {:?}", buf);
            return Err(alloy_rlp::Error::Custom("Invalid IPv4 address"));
        };
        let port = u16::decode(buf)?;
        let addr = SocketAddrV4::new(Ipv4Addr::from(ip), port);
        let seq = u64::decode(buf)?;

        Ok(Self { address: addr, seq })
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct MonadNameRecord<ST: CertificateSignatureRecoverable> {
    pub name_record: NameRecord,
    pub signature: ST,
}

impl<ST: CertificateSignatureRecoverable> MonadNameRecord<ST> {
    pub fn recover_pubkey(
        &self,
    ) -> Result<NodeId<CertificateSignaturePubKey<ST>>, <ST as CertificateSignature>::Error> {
        let mut encoded = Vec::new();
        self.name_record.encode(&mut encoded);
        let pubkey = self.signature.recover_pubkey(&encoded)?;
        Ok(NodeId::new(pubkey))
    }

    pub fn address(&self) -> SocketAddrV4 {
        self.name_record.address
    }

    pub fn seq(&self) -> u64 {
        self.name_record.seq
    }
}

pub enum PeerDiscoveryEvent<ST: CertificateSignatureRecoverable> {
    SendPing {
        to: NodeId<CertificateSignaturePubKey<ST>>,
    },
    PingRequest {
        from: NodeId<CertificateSignaturePubKey<ST>>,
        ping: Ping,
    },
    PongResponse {
        from: NodeId<CertificateSignaturePubKey<ST>>,
        pong: Pong,
    },
    PingTimeout {
        to: NodeId<CertificateSignaturePubKey<ST>>,
        ping_id: u32,
    },
    SendPeerLookup {
        to: NodeId<CertificateSignaturePubKey<ST>>,
        target: NodeId<CertificateSignaturePubKey<ST>>,
    },
    PeerLookupRequest {
        from: NodeId<CertificateSignaturePubKey<ST>>,
        request: PeerLookupRequest<ST>,
    },
    PeerLookupResponse {
        from: NodeId<CertificateSignaturePubKey<ST>>,
        response: PeerLookupResponse<ST>,
    },
    PeerLookupTimeout {
        to: NodeId<CertificateSignaturePubKey<ST>>,
        target: NodeId<CertificateSignaturePubKey<ST>>,
        lookup_id: u32,
    },
    Prune,
}

#[derive(Debug, Clone)]
pub enum PeerDiscoveryMessage<ST: CertificateSignatureRecoverable> {
    Ping(Ping),
    Pong(Pong),
    PeerLookupRequest(PeerLookupRequest<ST>),
    PeerLookupResponse(PeerLookupResponse<ST>),
}

impl<ST: CertificateSignatureRecoverable> Message for PeerDiscoveryMessage<ST> {
    type NodeIdPubKey = CertificateSignaturePubKey<ST>;
    type Event = PeerDiscoveryEvent<ST>;

    fn event(self, from: NodeId<Self::NodeIdPubKey>) -> Self::Event {
        match self {
            PeerDiscoveryMessage::Ping(ping) => PeerDiscoveryEvent::PingRequest { from, ping },
            PeerDiscoveryMessage::Pong(pong) => PeerDiscoveryEvent::PongResponse { from, pong },
            PeerDiscoveryMessage::PeerLookupRequest(request) => {
                PeerDiscoveryEvent::PeerLookupRequest { from, request }
            }
            PeerDiscoveryMessage::PeerLookupResponse(response) => {
                PeerDiscoveryEvent::PeerLookupResponse { from, response }
            }
        }
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum TimerKind {
    SendPing,
    PingTimeout,
    RetryPeerLookup { lookup_id: u32 },
    Prune,
}

pub enum PeerDiscoveryTimerCommand<E, ST: CertificateSignatureRecoverable> {
    Schedule {
        node_id: NodeId<CertificateSignaturePubKey<ST>>,
        timer_kind: TimerKind,
        duration: Duration,
        on_timeout: E,
    },
    ScheduleReset {
        node_id: NodeId<CertificateSignaturePubKey<ST>>,
        timer_kind: TimerKind,
    },
}

pub enum PeerDiscoveryCommand<ST: CertificateSignatureRecoverable> {
    RouterCommand {
        target: NodeId<CertificateSignaturePubKey<ST>>,
        message: PeerDiscoveryMessage<ST>,
    },
    TimerCommand(PeerDiscoveryTimerCommand<PeerDiscoveryEvent<ST>, ST>),
}

pub trait PeerDiscoveryAlgo {
    type SignatureType: CertificateSignatureRecoverable;

    fn send_ping(
        &mut self,
        target: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn handle_ping(
        &mut self,
        from: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
        ping: Ping,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn handle_pong(
        &mut self,
        from: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
        pong: Pong,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn handle_ping_timeout(
        &mut self,
        to: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
        ping_id: u32,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn send_peer_lookup_request(
        &mut self,
        to: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
        target: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn handle_peer_lookup_request(
        &mut self,
        from: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
        request: PeerLookupRequest<Self::SignatureType>,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn handle_peer_lookup_response(
        &mut self,
        from: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
        response: PeerLookupResponse<Self::SignatureType>,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn handle_peer_lookup_timeout(
        &mut self,
        to: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
        target: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
        lookup_id: u32,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn prune(&mut self) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn metrics(&self) -> &PeerDiscMetrics;

    fn get_sock_addr_by_id(
        &self,
        id: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
    ) -> Option<SocketAddrV4>;
}

pub trait PeerDiscoveryAlgoBuilder {
    type PeerDiscoveryAlgoType: PeerDiscoveryAlgo;

    fn build(
        self,
    ) -> (
        Self::PeerDiscoveryAlgoType,
        Vec<
            PeerDiscoveryCommand<<Self::PeerDiscoveryAlgoType as PeerDiscoveryAlgo>::SignatureType>,
        >,
    );
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use super::*;

    #[test]
    fn test_name_record_v4_rlp() {
        let name_record = NameRecord {
            address: SocketAddrV4::from_str("1.1.1.1:8000").unwrap(),
            seq: 2,
        };

        let mut encoded = Vec::new();
        name_record.encode(&mut encoded);

        let result = NameRecord::decode(&mut encoded.as_slice());
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), name_record);
    }
}
