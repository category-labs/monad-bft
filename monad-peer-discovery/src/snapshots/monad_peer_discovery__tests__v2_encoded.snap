---
source: monad-peer-discovery/src/lib.rs
expression: "hex::encode(&v2_encoded)"
---
"d284c0a83264cac480822328c4018223290f2a"
